File: cpp/lib/pairs.cpp
#include <algorithm>

#include "pairs.hpp"
#include "../util/display.hpp"

namespace Blastn {

#define ABS(value) ((s32)(value) < 0 ? (value) * -1 : (value))

vector<AdjacentPair> pair(vector<MatchSingleton>& flattened, u32 query_len)
{
    vector<AdjacentPair> result;

    for (u32 i = 0; i < flattened.size(); i++) {
        for (u32 j = i + 1; j < flattened.size(); j++) {
            // not overlapping
            if (ABS(((s32)flattened[i].sindex - (s32)flattened[j].sindex)) >= (s32)flattened[i].word.size()
                // not too far apart
                && ABS(((s32)flattened[i].sindex - (s32)flattened[j].sindex)) <= (s32)query_len - (s32)flattened[i].word.size())
            {
                result.emplace_back(
                    flattened[i].word,   flattened[j].word,
                    flattened[i].sindex, flattened[i].qindex,
                    flattened[j].sindex, flattened[j].qindex
                );

                break;
            }
        }
    }
    return result;
}

vector<AdjacentPair> flatten(vector<Match>& matches, u32 query_len)
{
    vector<MatchSingleton> flattened;

    for (auto& match : matches) {
        for (auto& sindex : match.subject_indices) {
            for (auto& qindex : match.query_indices) {
                flattened.push_back(MatchSingleton {
                    match.word, sindex, qindex
                });
            }
        }
    }
    std::sort(flattened.begin(), flattened.end(), [&](const MatchSingleton lhs, const MatchSingleton rhs) {
        return lhs.sindex < rhs.sindex;
    });

    return pair(flattened, query_len);
}

PairedSequenceMap pair_filter(MatchedSequenceMap& matches, SequenceMap& query)
{
    PairedSequenceMap filtered_pairs;
    Progress progress{ matches.size() };

    bool found = false;

    for (auto& sname_queries : matches) {
        PairedMatchesMap pairs;
        for (auto& qname_matches : sname_queries.second) {
            // qname not created yet in pairs
            for (auto& pair : flatten(qname_matches.second, (u32)query[qname_matches.first].size())) {
                if (ABS((s32)pair.sindex1 - (s32)pair.sindex2) <= (s32)query[qname_matches.first].size() - (s32)pair.length
                    || ABS((s32)pair.qindex1 - (s32)pair.qindex2) >= (s32)pair.length
                    || ABS((s32)pair.sindex1 - (s32)pair.sindex2) >= (s32)pair.length)
                {
                    // check if the pair was recorded yet
                    found = false;
                    for (auto& p : pairs[qname_matches.first]) {
                        if (pair.sindex1 == p.sindex1) {
                            found = true;
                            break;
                        }
                    }
                    // vector<Pair> already exists
                    if (!found)
                        pairs[qname_matches.first].emplace_back(pair);
                }
            }
        }
        if (!pairs.empty())
            filtered_pairs[sname_queries.first] = pairs;

        progress.update();
    }
    return filtered_pairs;
}

} // Blastn

File: python/lib/pairs.py
from typing import Dict, List, DefaultDict
import tqdm
from collections import defaultdict
from .match import MatchStruct

"""
Internal
"""
class MatchSingleton:
    def __init__(self, word: str, dindex: int, qindex: int):
        """
        @brief: Hold both words from the pair, how long they are, and where they both are in the query and database.
        """
        self.word: str = word
        self.dindex: int = dindex
        self.qindex: int = qindex

    def __str__(self):
        return str(self.__dict__)
    
    def __repr__(self):
        return self.__str__()


class AdjacentPair:
    def __init__(self, word1: str, word2: str, dindex1: int, qindex1: int, dindex2: int, qindex2: int):
        """
        @brief: Hold both words from the pair, how long they are, and where they both are in the query and database.
        """
        self.word1: str = word1
        self.word2: str = word2
        self.length = len(self.word1)
        self.dindex1: int = dindex1
        self.qindex1: int = qindex1
        self.dindex2: int = dindex2
        self.qindex2: int = qindex2

    def __str__(self):
        return str(self.__dict__)
    
    def __repr__(self):
        return self.__str__()

def append(flattened: List[MatchSingleton], result: list, query_len: int):
    for index1, value1 in enumerate(flattened):
        for index2 in range (index1 + 1, len(flattened)):
            if abs(value1.dindex - flattened[index2].dindex) >= len(value1.word) \
                and abs(value1.dindex - flattened[index2].dindex) <= query_len - len(value1.word):
                
                result.append(AdjacentPair(word1   = value1.word,       word2   = flattened[index2].word,
                                           dindex1 = value1.dindex, qindex1 = value1.qindex,
                                           dindex2 = flattened[index2].dindex, qindex2 = flattened[index2].qindex))
                break

def flatten(match_structs: List[MatchStruct], query_len: int):
    flat = []
    result: List[AdjacentPair] = []
    for match in match_structs:
        for dindex in match.data_indices:
            for qindex in match.query_indices:
                flat.append(MatchSingleton(match.word, dindex, qindex))
    flat.sort(key=lambda m: m.dindex)
    append(flat, result, query_len)
    
    return result

"""
External
"""

def pair_filter(matches: Dict[str, Dict[str, List[MatchStruct]]],
                query: Dict[str, str]) -> Dict[str, Dict[str, List[AdjacentPair]]]:

    filtered_pairs: Dict[str, Dict[str, List[AdjacentPair]]] = {}
    
    for dname, queries in tqdm.tqdm(matches.items()):
        pairs = defaultdict(list)
        for qname, match_structs in queries.items():
            for pair in flatten(match_structs, len(query[qname])):
                #                                    distance threshold
                if abs(pair.dindex1 - pair.dindex2) <= len(query[qname]) - pair.length \
                    or abs(pair.qindex1 - pair.qindex2) >= pair.length \
                    or abs(pair.dindex1 - pair.dindex2) >= pair.length:
                    # only append if the pair isn't already recorded
                    if not any(pair.dindex1 == p.dindex1 for p in pairs[qname]):
                        pairs[qname].append(pair)
        if pairs:
            filtered_pairs[dname] = dict(pairs)
    return filtered_pairs
